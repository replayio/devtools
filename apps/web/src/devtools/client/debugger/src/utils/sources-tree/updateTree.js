/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

//

import { addToTree } from "./addToTree";
import { collapseTree } from "./collapseTree";
import { createDirectoryNode, createParentMap } from "./utils";
import { getDomain } from "./treeOrder";

function getSourcesToAdd(newSources, prevSources) {
  const sourcesToAdd = [];

  for (const sourceId in newSources) {
    const newSource = newSources[sourceId];
    const prevSource = prevSources ? prevSources[sourceId] : null;
    if (!prevSource) {
      sourcesToAdd.push(newSource);
    }
  }

  return sourcesToAdd;
}

export function createTree({ debuggeeUrl, sources }) {
  const uncollapsedTree = createDirectoryNode("root", "", []);

  return updateTree({
    debuggeeUrl,
    newSources: sources,
    prevSources: {},
    uncollapsedTree,
  });
}

export function updateTree({ newSources, prevSources, debuggeeUrl, uncollapsedTree }) {
  const debuggeeHost = getDomain(debuggeeUrl);

  const sourcesToAdd = getSourcesToAdd(Object.values(newSources), Object.values(prevSources));

  for (const source of sourcesToAdd) {
    addToTree(uncollapsedTree, source, debuggeeHost);
  }

  const newSourceTree = collapseTree(uncollapsedTree);

  return {
    uncollapsedTree,
    sourceTree: newSourceTree,
    parentMap: createParentMap(newSourceTree),
  };
}
