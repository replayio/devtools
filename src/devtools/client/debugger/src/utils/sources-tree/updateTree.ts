/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

//

import { addToTree } from "./addToTree";
import { collapseTree } from "./collapseTree";
import { createDirectoryNode, createParentMap } from "./utils";

import type { Source } from "../../reducers/sources";
import type { TreeNode, TreeDirectory } from "./types";

type SourcesMap = Record<string, Source>;

function getSourcesToAdd(newSources: SourcesMap, prevSources: SourcesMap) {
  const sourcesToAdd = [];

  for (const sourceId in newSources) {
    const newSource = newSources[sourceId];
    const prevSource = prevSources ? prevSources[sourceId] : null;
    if (!prevSource) {
      sourcesToAdd.push(newSource);
    }
  }

  return sourcesToAdd;
}

export function createTree({ sources }: { sources: SourcesMap }) {
  // This used to be used, but it's not anymore.
  const uncollapsedTree = createDirectoryNode("root", "", []);

  return updateTree({
    debuggeeUrl: "",
    newSources: sources,
    prevSources: {},
    uncollapsedTree,
  });
}

interface UpdateTreeArgs {
  newSources: SourcesMap;
  prevSources: SourcesMap;
  uncollapsedTree: TreeDirectory;
  debuggeeUrl: string;
  sourceTree?: TreeNode;
}

export function updateTree({ newSources, prevSources, uncollapsedTree }: UpdateTreeArgs) {
  const debuggeeHost = "";

  // @ts-expect-error This used to be nested records - somehow it still works?
  const sourcesToAdd = getSourcesToAdd(Object.values(newSources), Object.values(prevSources));

  for (const source of sourcesToAdd) {
    addToTree(uncollapsedTree, source, debuggeeHost!);
  }

  const newSourceTree = collapseTree(uncollapsedTree);

  return {
    uncollapsedTree,
    sourceTree: newSourceTree,
    parentMap: createParentMap(newSourceTree),
  };
}
