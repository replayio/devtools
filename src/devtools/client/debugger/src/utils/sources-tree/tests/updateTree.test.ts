import type { SourceDetails } from "ui/reducers/sources";

import { createTree, updateTree } from "../updateTree";

describe("updateTree", () => {
  /*
  URLs taken from a Redux CodeSandbox recording:
  https://app.replay.io/recording/reduxts-csb-2--be1a4568-6c84-474c-a9eb-3a0491a792d8
  
  These demonstrate some key aspects of the sources tree behavior:
  - Multiple top-level folders
  - An "(index)" entry that is _not_ in a folder
  - Some nested folders that will be "collapsed" together
  - Sorting folders before files in the tree
  - Sorting files and folders alphabetically
  - Handling unusual "protocols" like "webpack://" and "replay-content://"
  - Current behavior when there's a name collision with different query params

  This array is intentionally written with these items
  out of order, to show that the sorting works correctly.
  */
  const partialPaths = [
    "src/index.tsx",
    "src/App.tsx",
    "src/app/store.ts",
    "src/app/hooks.ts",
    "webpack://BrowserFS/webpack/bootstrap",
    "(index)",
    "node_modules/redux/es/redux.js",
    "https://codesandbox.io/static/js/sandbox.js",
    "src/features/counter/Counter.tsx",
    "src/features/counter/counterAPI.ts",
    "src/features2/todos/todos.ts?1",
    "src/features2/todos/todos.ts?2",
    "replay-content:///react-devtools-hook-script",
    "https://codesandbox.io/static/js/vendors.js",
    "src/features/counter/counterSlice.ts",
    "webpack:////home/circleci/codesandbox-client/node_modules/core-js/fn/array/includes.js",
    "node_modules/react-redux/es/index.js",
    "node_modules/react-redux/es/types.js",
  ];

  // Case where two duplicated sources with the same content hash result in
  // a single "source" treenode
  const duplicatedSourcePaths1 = [
    "src/features3/todos/todos.ts?3",
    "src/features3/todos/todos.ts?4",
  ];

  // Case where there are multiple different source versions, but at least one
  // duplicate content hash, resulting in a "multiSource" node with <N items
  const duplicatedSourcePaths2 = [
    "src/features4/todos/todos.ts?number=5&duplicate=true",
    "src/features4/todos/todos.ts?number=6&duplicate=false",
    "src/features4/todos/todos.ts?number=7&duplicate=true",
  ];

  // Anything that starts with "webpack", "https", or "replay-content" is a full URL
  const reIsFullURL = /^(webpack|https|replay-content)/;

  // Fake a set of SourceDetails objects based on these URLs
  const mainSources: SourceDetails[] = partialPaths.map(path => {
    // We don't care about most of the SourceDetails object
    // fields, but TS wants us to fill them in anyway.
    // Cast to skip them.
    return {
      id: path,
      url: reIsFullURL.test(path) ? path : `https://something.com/${path}`,
      contentId: path,
    } as SourceDetails;
  });

  const duplicatedSources1: SourceDetails[] = duplicatedSourcePaths1.map(path => {
    const url = `https://something.com/${path}`;
    const urlObj = new URL(url);
    urlObj.search = "";
    return {
      id: path,
      url,
      // Fake a content hash by using the URL minus the query params,
      // which will fake a content hash collision
      contentId: urlObj.href,
    } as SourceDetails;
  });

  const duplicatedSources2: SourceDetails[] = duplicatedSourcePaths2.map(path => {
    const url = `https://something.com/${path}`;
    const urlObj = new URL(url);
    urlObj.search = "";
    return {
      id: path,
      url,
      // Fake a content hash by using the URL minus the query params,
      // which will fake a content hash collision
      contentId: path.includes("duplicate=true") ? urlObj.href : path,
    } as SourceDetails;
  });

  const sources = mainSources.concat(duplicatedSources1, duplicatedSources2);

  // Turn them into a record keyed by URLs, as used in the real app
  const sourcesRecord = Object.fromEntries(sources.map(source => [source.url, source] as const));

  test("Generates a correctly sorted nested tree of sources and folders", () => {
    // This also calls `updateTree` and `collapseTree` internally
    const tree = createTree({ sources: sourcesRecord });

    // Use snapshots just to verify that nothing has changed.
    // This does exercise all the tree manipulation logic.
    expect(tree.uncollapsedTree).toMatchSnapshot();
    expect(tree.sourceTree).toMatchSnapshot();
  });
});
