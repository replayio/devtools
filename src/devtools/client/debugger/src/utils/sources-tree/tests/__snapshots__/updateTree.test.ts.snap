// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateTree Generates a correctly sorted nested tree of sources and folders 1`] = `
Object {
  "contents": Array [],
  "name": "root",
  "path": "",
  "type": "directory",
}
`;

exports[`updateTree Generates a correctly sorted nested tree of sources and folders 2`] = `
Object {
  "contents": Array [
    Object {
      "contents": Array [
        Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "contents": Object {
                    "contentId": "https://codesandbox.io/static/js/sandbox.js",
                    "id": "https://codesandbox.io/static/js/sandbox.js",
                    "url": "https://codesandbox.io/static/js/sandbox.js",
                  },
                  "name": "sandbox.js",
                  "path": "/codesandbox.io/static/js/sandbox.js",
                  "type": "source",
                },
                Object {
                  "contents": Object {
                    "contentId": "https://codesandbox.io/static/js/vendors.js",
                    "id": "https://codesandbox.io/static/js/vendors.js",
                    "url": "https://codesandbox.io/static/js/vendors.js",
                  },
                  "name": "vendors.js",
                  "path": "/codesandbox.io/static/js/vendors.js",
                  "type": "source",
                },
              ],
              "name": "js",
              "path": "/codesandbox.io/static/js",
              "type": "directory",
            },
          ],
          "name": "static",
          "path": "/codesandbox.io/static",
          "type": "directory",
        },
      ],
      "name": "codesandbox.io",
      "path": "/codesandbox.io",
      "type": "directory",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Object {
            "contentId": "replay-content:///react-devtools-hook-script",
            "id": "replay-content:///react-devtools-hook-script",
            "url": "replay-content:///react-devtools-hook-script",
          },
          "name": "react-devtools-hook-script",
          "path": "/replay-content:///react-devtools-hook-script",
          "type": "source",
        },
      ],
      "name": "replay-content://",
      "path": "/replay-content://",
      "type": "directory",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Object {
            "contentId": "(index)",
            "id": "(index)",
            "url": "https://something.com/(index)",
          },
          "name": "(index)",
          "path": "/something.com/(index)",
          "type": "source",
        },
        Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "contents": Object {
                    "contentId": "node_modules/react-redux/es/index.js",
                    "id": "node_modules/react-redux/es/index.js",
                    "url": "https://something.com/node_modules/react-redux/es/index.js",
                  },
                  "name": "index.js",
                  "path": "/something.com/node_modules/react-redux/es/index.js",
                  "type": "source",
                },
                Object {
                  "contents": Object {
                    "contentId": "node_modules/react-redux/es/types.js",
                    "id": "node_modules/react-redux/es/types.js",
                    "url": "https://something.com/node_modules/react-redux/es/types.js",
                  },
                  "name": "types.js",
                  "path": "/something.com/node_modules/react-redux/es/types.js",
                  "type": "source",
                },
              ],
              "name": "react-redux/es",
              "path": "/something.com/node_modules/react-redux/es",
              "type": "directory",
            },
            Object {
              "contents": Array [
                Object {
                  "contents": Object {
                    "contentId": "node_modules/redux/es/redux.js",
                    "id": "node_modules/redux/es/redux.js",
                    "url": "https://something.com/node_modules/redux/es/redux.js",
                  },
                  "name": "redux.js",
                  "path": "/something.com/node_modules/redux/es/redux.js",
                  "type": "source",
                },
              ],
              "name": "redux/es",
              "path": "/something.com/node_modules/redux/es",
              "type": "directory",
            },
          ],
          "name": "node_modules",
          "path": "/something.com/node_modules",
          "type": "directory",
        },
        Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "contents": Object {
                    "contentId": "src/app/hooks.ts",
                    "id": "src/app/hooks.ts",
                    "url": "https://something.com/src/app/hooks.ts",
                  },
                  "name": "hooks.ts",
                  "path": "/something.com/src/app/hooks.ts",
                  "type": "source",
                },
                Object {
                  "contents": Object {
                    "contentId": "src/app/store.ts",
                    "id": "src/app/store.ts",
                    "url": "https://something.com/src/app/store.ts",
                  },
                  "name": "store.ts",
                  "path": "/something.com/src/app/store.ts",
                  "type": "source",
                },
              ],
              "name": "app",
              "path": "/something.com/src/app",
              "type": "directory",
            },
            Object {
              "contents": Array [
                Object {
                  "contents": Object {
                    "contentId": "src/features/counter/Counter.tsx",
                    "id": "src/features/counter/Counter.tsx",
                    "url": "https://something.com/src/features/counter/Counter.tsx",
                  },
                  "name": "Counter.tsx",
                  "path": "/something.com/src/features/counter/Counter.tsx",
                  "type": "source",
                },
                Object {
                  "contents": Object {
                    "contentId": "src/features/counter/counterAPI.ts",
                    "id": "src/features/counter/counterAPI.ts",
                    "url": "https://something.com/src/features/counter/counterAPI.ts",
                  },
                  "name": "counterAPI.ts",
                  "path": "/something.com/src/features/counter/counterAPI.ts",
                  "type": "source",
                },
                Object {
                  "contents": Object {
                    "contentId": "src/features/counter/counterSlice.ts",
                    "id": "src/features/counter/counterSlice.ts",
                    "url": "https://something.com/src/features/counter/counterSlice.ts",
                  },
                  "name": "counterSlice.ts",
                  "path": "/something.com/src/features/counter/counterSlice.ts",
                  "type": "source",
                },
              ],
              "name": "features/counter",
              "path": "/something.com/src/features/counter",
              "type": "directory",
            },
            Object {
              "contents": Array [
                Object {
                  "contents": Array [
                    Object {
                      "contents": Object {
                        "contentId": "src/features2/todos/todos.ts?1",
                        "id": "src/features2/todos/todos.ts?1",
                        "url": "https://something.com/src/features2/todos/todos.ts?1",
                      },
                      "name": "(1) todos.ts",
                      "path": "https://something.com/src/features2/todos/todos.ts?1",
                      "type": "source",
                    },
                    Object {
                      "contents": Object {
                        "contentId": "src/features2/todos/todos.ts?2",
                        "id": "src/features2/todos/todos.ts?2",
                        "url": "https://something.com/src/features2/todos/todos.ts?2",
                      },
                      "name": "(2) todos.ts",
                      "path": "https://something.com/src/features2/todos/todos.ts?2",
                      "type": "source",
                    },
                  ],
                  "name": "todos.ts",
                  "path": "/something.com/src/features2/todos/todos.ts",
                  "type": "multiSource",
                },
              ],
              "name": "features2/todos",
              "path": "/something.com/src/features2/todos",
              "type": "directory",
            },
            Object {
              "contents": Array [
                Object {
                  "contents": Object {
                    "contentId": "https://something.com/src/features3/todos/todos.ts",
                    "id": "src/features3/todos/todos.ts?3",
                    "url": "https://something.com/src/features3/todos/todos.ts?3",
                  },
                  "name": "todos.ts",
                  "path": "/something.com/src/features3/todos/todos.ts",
                  "type": "source",
                },
              ],
              "name": "features3/todos",
              "path": "/something.com/src/features3/todos",
              "type": "directory",
            },
            Object {
              "contents": Array [
                Object {
                  "contents": Array [
                    Object {
                      "contents": Object {
                        "contentId": "https://something.com/src/features4/todos/todos.ts",
                        "id": "src/features4/todos/todos.ts?number=5&duplicate=true",
                        "url": "https://something.com/src/features4/todos/todos.ts?number=5&duplicate=true",
                      },
                      "name": "(1) todos.ts",
                      "path": "https://something.com/src/features4/todos/todos.ts?number=5&duplicate=true",
                      "type": "source",
                    },
                    Object {
                      "contents": Object {
                        "contentId": "src/features4/todos/todos.ts?number=6&duplicate=false",
                        "id": "src/features4/todos/todos.ts?number=6&duplicate=false",
                        "url": "https://something.com/src/features4/todos/todos.ts?number=6&duplicate=false",
                      },
                      "name": "(2) todos.ts",
                      "path": "https://something.com/src/features4/todos/todos.ts?number=6&duplicate=false",
                      "type": "source",
                    },
                  ],
                  "name": "todos.ts",
                  "path": "/something.com/src/features4/todos/todos.ts",
                  "type": "multiSource",
                },
              ],
              "name": "features4/todos",
              "path": "/something.com/src/features4/todos",
              "type": "directory",
            },
            Object {
              "contents": Object {
                "contentId": "src/App.tsx",
                "id": "src/App.tsx",
                "url": "https://something.com/src/App.tsx",
              },
              "name": "App.tsx",
              "path": "/something.com/src/App.tsx",
              "type": "source",
            },
            Object {
              "contents": Object {
                "contentId": "src/index.tsx",
                "id": "src/index.tsx",
                "url": "https://something.com/src/index.tsx",
              },
              "name": "index.tsx",
              "path": "/something.com/src/index.tsx",
              "type": "source",
            },
          ],
          "name": "src",
          "path": "/something.com/src",
          "type": "directory",
        },
      ],
      "name": "something.com",
      "path": "/something.com",
      "type": "directory",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "contents": Object {
                    "contentId": "webpack:////home/circleci/codesandbox-client/node_modules/core-js/fn/array/includes.js",
                    "id": "webpack:////home/circleci/codesandbox-client/node_modules/core-js/fn/array/includes.js",
                    "url": "webpack:////home/circleci/codesandbox-client/node_modules/core-js/fn/array/includes.js",
                  },
                  "name": "includes.js",
                  "path": "/webpack:////home/circleci/codesandbox-client/node_modules/core-js/fn/array/includes.js",
                  "type": "source",
                },
              ],
              "name": "home/circleci/codesandbox-client/node_modules/core-js/fn/array",
              "path": "/webpack:////home/circleci/codesandbox-client/node_modules/core-js/fn/array",
              "type": "directory",
            },
          ],
          "name": "",
          "path": "/webpack:///",
          "type": "directory",
        },
        Object {
          "contents": Array [
            Object {
              "contents": Object {
                "contentId": "webpack://BrowserFS/webpack/bootstrap",
                "id": "webpack://BrowserFS/webpack/bootstrap",
                "url": "webpack://BrowserFS/webpack/bootstrap",
              },
              "name": "bootstrap",
              "path": "/webpack:///webpack/bootstrap",
              "type": "source",
            },
          ],
          "name": "webpack",
          "path": "/webpack:///webpack",
          "type": "directory",
        },
      ],
      "name": "webpack://",
      "path": "/webpack://",
      "type": "directory",
    },
  ],
  "name": "root",
  "path": "",
  "type": "directory",
}
`;
